src/profile-logic/comparison.js

```diff
@@ -246,3 +309,498 @@ function adjustNullableCategories(
+          This is a programming error.
+        `);
+      }
+
+      const nameIndex = resourceTable.name[i];
+      const newName = nameIndex >= 0 ? stringTable.getString(nameIndex) : null;
+
+      const hostIndex = resourceTable.host[i];
+      const newHost =
+        typeof hostIndex === 'number' && hostIndex >= 0
+          ? stringTable.getString(hostIndex)
+          : undefined;
+
+      const type = resourceTable.type[i];
+
+      // Duplicate search.
+      const resourceKey = [newLibIndex, newName || '', type].join('#');
+      const insertedResourceIndex = mapOfInsertedResources.get(resourceKey);
+      if (insertedResourceIndex !== undefined) {
+        translationMap.set(i, insertedResourceIndex);
+        continue;
+      }
+
+      translationMap.set(i, newResourceTable.length);
+      mapOfInsertedResources.set(resourceKey, newResourceTable.length);
+
+      newResourceTable.lib.push(newLibIndex);
+      newResourceTable.name.push(
+        newName === null ? -1 : newStringTable.indexForString(newName)
+      );
+      newResourceTable.host.push(
+        newHost === undefined
+          ? undefined
+          : newStringTable.indexForString(newHost)
+      );
+      newResourceTable.type.push(type);
+
+      newResourceTable.length++;
+    }
+
+    translationMaps.push(translationMap);
+  });
+
+  return { resourceTable: newResourceTable, translationMaps };
+}
+
+/**
+ * This combines the function tables for a list of threads. It returns the new
+ * function table with the translation maps to be used in subsequent merging
+ * functions.
+ */
+function combineFuncTables(
+  translationMapsForResources: TranslationMapForResources[],
+  newStringTable: UniqueStringArray,
+  ...threads: Thread[]
+): { funcTable: FuncTable, translationMaps: TranslationMapForFuncs[] } {
+  const mapOfInsertedFuncs: Map<string, IndexIntoFuncTable> = new Map();
+  const translationMaps = [];
+  const newFuncTable = getEmptyFuncTable();
+
+  threads.forEach((thread, threadIndex) => {
+    const { funcTable, stringTable } = thread;
+    const translationMap = new Map();
+    const resourceTranslationMap = translationMapsForResources[threadIndex];
+
+    for (let i = 0; i < funcTable.length; i++) {
+      const fileNameIndex = funcTable.fileName[i];
+      const fileName =
+        typeof fileNameIndex === 'number'
+          ? stringTable.getString(fileNameIndex)
+          : null;
+      const resourceIndex = funcTable.resource[i];
+      const newResourceIndex =
+        resourceIndex >= 0
+          ? resourceTranslationMap.get(funcTable.resource[i])
+          : -1;
+      if (newResourceIndex === undefined) {
+        throw new Error(stripIndent`
+          We couldn't find the resource of func ${i} in the translation map.
+          This is a programming error.
+        `);
+      }
+      const name = stringTable.getString(funcTable.name[i]);
+
+      const funcKey = [name, newResourceIndex].join('#');
```

Comment:
Is this unique enough of a key? What about JS, can we have the same function name multiple times?

Reply:
Good question. I don't know how we can distinguish these :/
That said, we're mostly interested in the stacks, and I think that would still work. Unless the user chooses to "focus on function"...

Reply:
I think it would be good to do a bit of due diligence to figure out how to break the uniqueness here, as this could be a source of "lying to the user" in the future. I think from a JS perspective, we could probably be unique on the combination of func name, line number, and script source. I'm not sure on the C++ side where the uniqueness comes in. I know the original mangled symbols are unique names. Maybe Gerald or Markus would be good to ask.

Reply:
yep sure, I'll dig this deeper :)

Reply:
For c++ I think that using the resourceIndex makes it unique. For JS I'll use the line number as well indeed (the file name should be included in the resource already), as 2 functions might have the same name in the same file.

Reply:
Note: for cpp looks like that in the process phase we don't even look at the resource index: https://github.com/firefox-devtools/profiler/blob/ebd5b869c5a3e0ff30b97b7ffc3fd069df1cf2d7/src/profile-logic/process-profile.js#L322
But this isn't the same context, as at the processing phase we know that everything is from the same binary.

In the future we may relax this a bit so that we can compare eg Mac vs Linux, but this would be challenging too because of stack differences anyway...